File: companies/Deutsche Bank/82.remove-duplicates-from-sorted-list-ii.java
/**
 * Algorithm:
 * Since the list is sorted, duplicates occur in consecutive runs. A dummy node
 * handles duplicated runs at the head. prev points to the last confirmed unique
 * node; cur scans each run and either keeps or removes it.
 *
 * Complexity:
 * Time O(n), space O(1).
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        ListNode cur = head;
        while (cur != null) {
            boolean duplicate = false;
            while (cur.next != null && cur.val == cur.next.val) {
                duplicate = true;
                cur = cur.next;
            }
            if (duplicate) {
                prev.next = cur.next;
            } else {
                prev = prev.next;
            }
            cur = cur.next;
        }
        return dummy.next;
    }
}

File: data structure and algorithms/Two Pointers/82.remove-duplicates-from-sorted-list-ii.go
package leetcode

// DeleteDuplicates82 removes every duplicated run from a sorted list. A dummy
// node handles duplicate runs at the head; prev stays at the last confirmed
// unique node while cur scans each run.
//
// Time: O(n). Space: O(1).
func DeleteDuplicates82(head *ListNode) *ListNode {
	dummy := &ListNode{Next: head}
	prev := dummy
	cur := head
	for cur != nil {
		dup := false
		for cur.Next != nil && cur.Val == cur.Next.Val {
			dup = true
			cur = cur.Next
		}
		if dup {
			prev.Next = cur.Next
		} else {
			prev = prev.Next
		}
		cur = cur.Next
	}
	return dummy.Next
}
